### PR TITLE
tx: Generic Transaction Interface

### PR DESCRIPTION
**Pull request by gabrocheleau**
Resolves issue #2726 by implementing a generic Transaction interface and additional types/interfaces for related ones (e.g. `TxData`, `TxValuesArray`)

### Notable changes

- Renamed `Transaction` to `LegacyTransaction`. There were some comments in the repo that the `Transaction` naming was kept for retro-compatibility, so not sure if we can allow ourselves to rename it for this breaking release.
- Added a `TransactionType` enum for the 4 transaction types
- Added a `TransactionInterface` (similar to the state manager interface) that specifies the methods that should be implement by all transaction classes & uses generics to significantly improve type inference. 
- The `BaseTransaction` abstract class now implements `TransactionInterface` 
- Added a `Transaction[TransactionType], TxData[Transaction[TransactionType], TxValuesArray[TransactionType]` interfaces that can be used to map a transaction type to its corresponding data types.
- Added TypeGuard methods to verify if Tx/TxData belongs to a certain TxType (e.g. `isLegacyTx(txData: TypedTransaction): tx is LegacyTransaction`)

### CODE REVIEW

File: packages/tx/src/baseTransaction.ts
Comment (g11tech) [inline packages/tx/src/baseTransaction.ts:46]:
```suggestion
export abstract class BaseTransaction<T extends TransactionType>
```
just using `T` should be sufficient and understandable for generics while keeping it short

File: packages/tx/src/eip2930Transaction.ts
Comment (g11tech) [inline packages/tx/src/eip2930Transaction.ts:148]: I would like to define all ` ...[TransactionType.AccessListEIP2930]` references on top of each transaction class file as:
```
type TxData = AllTypesTxData[TransactionType.AccessListEIP2930]
```
and then use just `TxData`, `TxValuesArray` etc etc in the class def for brevity

File: packages/tx/src/types.ts
Comment (g11tech) [inline packages/tx/src/types.ts:154]: may be you can do this: `TxData[TransactionType]` and should give the same meaning

also would not like to call this as `UnknownTxData` but `TypedTransaction` was a better name imo

File: packages/tx/src/types.ts
Comment (g11tech) [inline packages/tx/src/types.ts:147]: like this idea very much :heart:

File: packages/tx/src/types.ts
Comment (g11tech) [inline packages/tx/src/types.ts:170]: very nice typeguards :clap:

File: packages/tx/src/types.ts
Comment (g11tech) [inline packages/tx/src/types.ts:168]: Instead of 1,2,3 num comparisions, can do we enum `TransactionType` in a proper way assinging the type numbers to it explicitly and then use the enum for comparision.

File: packages/tx/src/types.ts
Comment (g11tech) [inline packages/tx/src/types.ts:119]: i think the `TypedTransaction[]` is more meaningful and code intuitive than `TransactionsArray` and would like to retain its usage even in the new refactored form

File: packages/tx/src/types.ts
Comment (gabrocheleau) [reply packages/tx/src/types.ts:168]: Good point, will do!

Comment (gabrocheleau) [reply packages/tx/src/types.ts]: Done! Also added additional typeguards and replaced a switch statement with that.

File: packages/tx/src/baseTransaction.ts
Comment (gabrocheleau) [reply packages/tx/src/baseTransaction.ts:46]: Good point, done.

File: packages/tx/src/types.ts
Comment (holgerd77) [reply packages/tx/src/types.ts:135]: Yes, would agree here.

Comment (gabrocheleau) [reply packages/tx/src/types.ts]: Renamed to TypedTransaction & TypedTransaction[]

File: packages/tx/src/types.ts
Comment (gabrocheleau) [reply packages/tx/src/types.ts:168]: Oh indeed, nice. Implemented and updated the naming to TypedTxData

File: packages/tx/src/eip2930Transaction.ts
Comment (gabrocheleau) [reply packages/tx/src/eip2930Transaction.ts]: Good idea, done!

File: packages/client/src/net/protocol/ethprotocol.ts
Comment (g11tech) [inline packages/client/src/net/protocol/ethprotocol.ts:234]: break here will break the for loop, continue is better , but overall it should be chained in if else if...

Comment (g11tech) [reply packages/client/src/net/protocol/ethprotocol.ts]: updated

**Review by g11tech — APPROVED** [review]

File: packages/client/src/net/protocol/ethprotocol.ts
Comment (gabrocheleau) [reply packages/client/src/net/protocol/ethprotocol.ts]: Thanks!

**Review by g11tech — APPROVED** [review]

**Comment by gabrocheleau** [comment]
Thanks for approving @g11tech! I'll wait for @holgerd77 's final approval to merge to make sure this PR properly addresses the issue he had created. (#2726)

File: packages/tx/src/eip1559Transaction.ts
Comment (holgerd77) [inline packages/tx/src/eip1559Transaction.ts:60]: Uhg.

We still have got "London" in here? 🤔

So this is definitely a left over. Just had a look to (re-)get the whole picture, some context: we introduced this "sub"-default HF for tx types which are not yet activated on the current broader default HF (so historically these were actually not "back" but "forward" 😋).

So for 1559 (London) and 2930 (Berlin) we should definitely remove completely.

We can eventually think if we rather want to add for 4844 and set to "Cancun" (so *that* was actually the original intended use case 😋).

On thinking about I would rather think we do not want to do though, seems to have a tendency to get forgotten and think we should be fine with the concrete ´this.common.isActivatedEIP(4844)` capability/EIP checks in the constructor (which is more tx-type-adequate anyhow).

Nevertheless, so: we should remove these two sub-default-HFs before breaking releases!

File: packages/tx/src/types.ts
Comment (holgerd77) [inline packages/tx/src/types.ts:137]: Ah, we should urgently have a look there and see that we align this typing to a single version, pretty sure (not 100% though, so this would need some confirmation) these overload-typings have been introduced for some backwards-compatibility reasons.

File: packages/tx/src/types.ts
Comment (holgerd77) [inline packages/tx/src/types.ts:145]: Same here.

If I remember correctly there should be some valuable and extensive additional context in the comments for these methods.

**Review by holgerd77 — APPROVED** [review]
🤯

Oh oh oh.

I honestly didn't expect something so wonderful come out of this task in such an ad hoc manner!!! 🎉

Thanks so so much, all these simplifications are a delight to look at, throughout the whole PR. 🙏

Two remarks, but rather very much unrelated and for a follow-up PR, either for you or someone else.

Also, just to confirm respectively to keep some confirmative discussion going a bit: so one of the larger goals for this task was that it I guess both gets easier a) for people in the tx library itself to implement their own tx type and as well b) *use* this tx type in our other libraries.

For a) the direct question: how much did you have this in mind and would you think the structure is better prepared for this now? In particular, is this tx type enum eventually somewhere in the way to do so, or is this well integrated/expandable?

For b) I guess we need to have some differentiated look anyhow: so for VM this is (at least for now and for the short-term future) very much "bound" regarding compatibility to the existing tx types (generally interesting discussion though if this *could* theoretically be generalized, will leave this here though, going too far).

So I guess we can start the compatibility question on Block - eventually Blockchain (but rather Block for now) - question. Is Block now (better) prepared to eventually take a custom tx type? 🤔

Still thinking a bit loud myself: we likely also *do not want* to have Block accept all sorts of types e.g. for a predefined chain (so e.g. for a mainnet Cancun chain accept Optimism system txs).

I wonder if would want to just explicitly add the accepted tx types to chain files in Common, so that this gets:

```json
{
  "name": "london",
  "block": 12965000,
  "txTypes": [0, 1, 2],
  "forkHash": "0xb715077d"
}
```

This would make things more consistent. Wouldn't be much to add (likely also more clean for our own checks) and then people could define their own chain files with additional type numbers, implement their tx type and block and blockchain validations can then likely (?) be written to check for this respective types in Common.

Anyhow: this is going a lot further and up for a separate discussion! 😀

Again, great work!!!

Will merge.

File: packages/tx/src/eip1559Transaction.ts
Comment (gabrocheleau) [reply packages/tx/src/eip1559Transaction.ts:60]: Sure, will push a PR for that!

**Comment by gabrocheleau** [comment]
> exploding_head
> 
> Oh oh oh.
> 
> I honestly didn't expect something so wonderful come out of this task in such an ad hoc manner!!! tada
> 
> Thanks so so much, all these simplifications are a delight to look at, throughout the whole PR. pray
> 
> Two remarks, but rather very much unrelated and for a follow-up PR, either for you or someone else.
> 
> Also, just to confirm respectively to keep some confirmative discussion going a bit: so one of the larger goals for this task was that it I guess both gets easier a) for people in the tx library itself to implement their own tx type and as well b) _use_ this tx type in our other libraries.
> 
> For a) the direct question: how much did you have this in mind and would you think the structure is better prepared for this now? In particular, is this tx type enum eventually somewhere in the way to do so, or is this well integrated/expandable?
> 
> For b) I guess we need to have some differentiated look anyhow: so for VM this is (at least for now and for the short-term future) very much "bound" regarding compatibility to the existing tx types (generally interesting discussion though if this _could_ theoretically be generalized, will leave this here though, going too far).
> 
> So I guess we can start the compatibility question on Block - eventually Blockchain (but rather Block for now) - question. Is Block now (better) prepared to eventually take a custom tx type? thinking
> 
> Still thinking a bit loud myself: we likely also _do not want_ to have Block accept all sorts of types e.g. for a predefined chain (so e.g. for a mainnet Cancun chain accept Optimism system txs).
> 
> I wonder if we would want to just explicitly add the accepted tx types to chain files in Common, so that this gets:
> 
> ```json
> {
>   "name": "london",
>   "block": 12965000,
>   "txTypes": [0, 1, 2],
>   "forkHash": "0xb715077d"
> }
> ```
> 
> This would make things more consistent. Wouldn't be much to add (likely also more clean for our own checks) and then people could define their own chain files with additional type numbers, implement their tx type and block and blockchain validations can then likely (?) be written to check for this respective types in Common.
> 
> Anyhow: this is going a lot further and up for a separate discussion! grinning
> 
> Again, great work!!!
> 
> Will merge.

Hey @holgerd77!

Thanks for the detailed comments and reflections!

Regarding a) and b) 

I did not directly focus on expandability with this PR, but rather more on refactoring and improving types. However, I think this PR still brings us closer to enabling custom a) (custom tx types). 

I'm not actually sure about the enum itself, makes me wonder if we could also add a TransactionType.Custom (?), although I'm not sure how helpful that would be. One nice thing about this updated structure is that a user that would add a 4th transaction type to the enum (e.g. `TransactionType.Optimism`) would then get a cascade of type issues throughout the repo allowing them to easily spot instances where some implementation is missing (e.g. switch statements that are missing the case for that 4th txtype). They could also easily create a new Transaction class that extends BaseTransaction, and therefore comes with the constraints of the TransactionInterface (just like when implementing a new StateManager). 

I'm not sure how much ease of extensibility we can realistically aim for with other packages though. As you imply in your comments regarding b), txs are quite a fundamental building block, and new tx types would I assume come with all sorts of idiosyncratic behavior that affect blocks, vm, etc. I feel like there is a limit to how much generalization we can do. As a thought experiment, pre-4844, how could we have generalized Txs in a way that allowed for 4844 transactions? Some of these are bound to be so "outside the box" that I can hardly see how we could generalize transactions in a way that's robust and future-proof.

Really interesting thing to reflect on though! We could eventually add an example of implementing a custom tx type, just like we do for some of the other packages.

**Pull request merged**
